### PR TITLE
Update to spec v0.6.1 (local branch)

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Update to v0.6.1 of the spec.

## Additional Info

This is the #368 PR, but on a branch local to this repo (not in @michaelsproul's repo).

## TODO:

- ~~Ensure `Deposit.signature` is only verified for new validators.~~

### CODE REVIEW

File: eth2/state_processing/src/per_block_processing.rs [hunk 1]
```diff
@@ -469,7 +469,7 @@ pub fn process_transfers<T: EthSpec>(
     spec: &ChainSpec,
 ) -> Result<(), Error> {
     verify!(
-        transfers.len() as u64 <= spec.max_transfers,
+        transfers.len() as u64 >= spec.max_transfers,
```
Comment (michaelsproul) [inline eth2/state_processing/src/per_block_processing.rs]: Did you mean to commit this?

Comment (paulhauner) [reply eth2/state_processing/src/per_block_processing.rs]: Oh I did not! Good catch, thanks

Comment (michaelsproul) [reply eth2/state_processing/src/per_block_processing.rs]: Np, saw it in my notifications and was like hmmm 🤔

**Comment by paulhauner** [comment]
@michaelsproul, I'm wondering about how we know when the `ExitCache` is not adequately initialized.

The other caches have the following properties regarding initialization:

- `CommitteeCache`: knows the epoch it's initialized for, returns errors based upon that.
- `PubkeyCache`: can determine if it's up to date based upon the length of the `validator_registry` and do an incremental catchup.
- `TreeHashCache`: doesn't really care if it's out-of-date it just does its thing.

A solution for `ExitCache` wasn't immediately obvious to me, but perhaps we can find something?

**Comment by michaelsproul** [comment]
The simplest thing I can think of doing to improve the consistency checks on the `ExitCache` would be to store the set of validators that are exiting at each epoch. Then `record_validator_exit` can at least enforce its precondition that it must only be called once for each exiting validator (it can check if the validator being exited is already in the set).

Other than that we could maybe have an `initialised: bool` on the `ExitCache` to at least ensure that it has been initialised to something other than `ExitCache::default`

File: eth2/state_processing/src/per_block_processing/verify_transfer.rs [hunk 2]
```diff
@@ -77,19 +95,33 @@ fn verify_transfer_parametric<T: EthSpec>(
...
     let sender_validator = state
         .validator_registry
         .get(transfer.sender as usize)
         .ok_or_else(|| Error::Invalid(Invalid::FromValidatorUnknown(transfer.sender)))?;
+
     let epoch = state.slot.epoch(spec.slots_per_epoch);
 
+    // Ensure one of the following is met:
+    //
+    // - Time independent checks are being ignored.
```
Comment (michaelsproul) [inline eth2/state_processing/src/per_block_processing/verify_transfer.rs]: I think this should be _Time dependent checks are being ignored._

Comment (michaelsproul) [reply eth2/state_processing/src/per_block_processing/verify_transfer.rs]: That said, the "time independence" concept isn't very well named, we could change the variable to something like `allow_future_compatible` (as in, treat as valid any transactions which could become valid in the future)

**Comment by paulhauner** [comment]
Hey @michaelsproul, sorry for the Sunday comms -- please ignore until Monday. As you may know, I've been running testnets and analyzing the metrics over on the `docker-env` branch. This branch is still using `v0.5.1` and I've hit some bugs that are blocking progress. I'm hesitant to spend any time on `v0.5.1` code, so I'm thinking the best thing to do is try and get this branch merged into my `docker-env` branch.

To achieve this, I think we should get this branch building in it's entirety (i.e., passing CI, but perhaps not all EF tests) and then I can merge it into my `docker-env` branch.

I merged `master` into this branch (feel free to revert/force-push if you dislike that) to see if I could get it to build. However, `operations_pool` isn't compiling and I'm not confident my fixes would be holistically sane.

I'm going to be on a plane tomorrow (Monday) but planes have WiFi these days so I'm thinking we can still work together. Unless you disagree, I'm thinking that if you can fix the op-pool branch in the morning, I can probably fix the rest of it. Once it's building, I merge it into my `docker-env` branch and continue testing.

Let me know your thoughts, I'll reach out once I have plan WiFI :)

**Comment by paulhauner** [comment]
I've merged my `docker-env` branch into this branch. It includes the following significant changes:

- We only have two `EthSpec`'s now and they both reflect the specs that are defined by the EF tests:
    - `MainnetEthSpec`: this replaces `FoundationEthSpec`
    - `MinimalEthSpec`: this is basically `FewValidatorsEthSpec`/`LighthouseTestnetEthSpec`.
- The `EthSpec::spec()` fn has been renamed to `EthSpec::default_spec()`. Because I increased the flexibility of spec config, we can't just pull `ChainSpec` from the `EthSpec` anymore because it might have been modified at runtime. There are still several occasions where we use `EthSpec::default_spec()` in testing, this is fine.
- There are three TOML configs across the project. They are generated automatically from `Default` impls if they don't exist and can have custom paths specified via CLI:
    - `eth2-spec.toml`: Contains spec parameters, basically `ChainSpec` serialized. Consistent across all binaries, generally there should be a single one of these as a source of truth.
    - `beacon-node.toml`: run-time config for the beacon node.
    - `validator-client.toml`: run-time config for validator client.
    - `account-manager.toml`: run-time config for the account manager.
- We have a new CLI flag for BN/VC: `--recent-genesis` (`-r`) that overrides the genesis time loaded from `eth2-spec.toml` with a time at the last 30 minute boundary (this used to be our default behavior).

There are also changes to metrics, syncing and block processing.

**Note:** 4f43a67 adds `#[should_panic]` to a few failing EF tests. I've done this so we can progress forwards to passing CI. @michaelsproul, this is relevant to you :)

**Comment by michaelsproul** [comment]
I think this PR is ready to review & merge! 

I have one outstanding question: what's the easiest way to update the `validator/eth2_config.toml` for the changes I made to the minimal spec?

**Comment by paulhauner** [comment]
@michaelsproul that's awesome! I'll start a review on my travels home.

The `eth2_config.toml` file is generated from the `ChainSpec` definitions, so it should update automatically. The [logic](https://github.com/sigp/lighthouse/blob/9cec5dc073ae4fa4608b787c1b624589add92fd0/validator_client/src/main.rs#L117-L138) is like this:
- If `eth2_config.toml` exists on disk, use it.
- Else, read the `spec-constants` flag from the CLI (`mainnet` or `minimal`) and load one of those, writing a new config to disk.

So effectively, just delete the one you have now and it'll generate a new one for you :)

**Pull request merged**